### PR TITLE
Report exponential_avg_bucket_processing_time which gives more weight to recent buckets

### PR DESCRIPTION
**Pull request by przemekwitek**
Report exponential_avg_bucket_processing_time which gives more weight to recent buckets.

Related issue: #29857

### CODE REVIEW

**Comment by droberts195** [comment]
Now the `timing_stats` object is complete it would be good to add it to the docs in `docs/reference/ml/apis/jobcounts.asciidoc`.

Also, it would be good to add an example `timing_stats` (that's believable given the other stats) into the example response in `docs/reference/ml/apis/get-job-stats.asciidoc`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/TimingStats.java [hunk 1]
```diff
@@ -107,12 +121,16 @@ public Double getAvgBucketProcessingTimeMs() {
...
+        return exponentialAvgBucketProcessingTimeMs;
+    }
+
     /**
      * Updates the statistics (min, max, avg) for the given data point (bucket processing time).
      */
     public void updateStats(double bucketProcessingTimeMs) {
         if (bucketProcessingTimeMs < 0.0) {
-            throw new IllegalArgumentException("bucketProcessingTimeMs must be positive, was: " + bucketProcessingTimeMs);
+            throw new IllegalArgumentException("bucketProcessingTimeMs must be non-negative, was: " + bucketProcessingTimeMs);
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/TimingStats.java]: #pedantry 
So, is `0.0` considered a negative number? 😂

**Review by benwtrent — COMMENTED** [review]
I think adding docs and updating the `yml` REST tests would be good to do in this PR as well. My only comment in the review is not that important :D

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/TimingStats.java (hunk 1 above)
Comment (przemekwitek) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/TimingStats.java]: I'm not sure if you've seen the latest revision of this comment ("bucketProcessingTimeMs must be non-negative, was: ") or not. But now I believe it is correct i.e. if the value is less than 0 we throw an error that it should be non-negative (so 0 or greater than 0).

**Comment by przemekwitek** [comment]
> Now the `timing_stats` object is complete it would be good to add it to the docs in `docs/reference/ml/apis/jobcounts.asciidoc`.

Done.

> Also, it would be good to add an example `timing_stats` (that's believable given the other stats) into the example response in `docs/reference/ml/apis/get-job-stats.asciidoc`.

Done.

**Comment by przemekwitek** [comment]
> I think adding docs and updating the `yml` REST tests would be good to do in this PR as well. My only comment in the review is not that important :D

Done.

File: docs/reference/ml/apis/jobcounts.asciidoc [hunk 2]
```diff
@@ -209,6 +213,31 @@ The `forecasts_stats` object shows statistics about forecasts. It has the follow
...
+  (long) The number of buckets processed.
+
+`minimum_bucket_processing_time_ms`::
+  (double) Minimum among all bucket processing times in milliseconds.
+
+`maximum_bucket_processing_time_ms`::
+  (double) Maximum among all bucket processing times in milliseconds.
+
+`average_bucket_processing_time_ms`::
+  (double) Cumulative moving average of all bucket processing times in milliseconds.
```
Comment (droberts195) [inline docs/reference/ml/apis/jobcounts.asciidoc]: Should this contain the word "moving"?

File: x-pack/plugin/src/test/resources/rest-api-spec/test/ml/jobs_get_stats.yml [hunk 3]
```diff
@@ -88,18 +89,22 @@ setup:
...
+  - is_true:  jobs.0.node.name
+  - is_true:  jobs.0.node.transport_address
+  - match:  { jobs.0.node.attributes.ml\.max_open_jobs: "20"}
+  - is_true:  jobs.0.open_time
+  - match:  { jobs.0.timing_stats.job_id: job-stats-test }
+  - match:  { jobs.0.timing_stats.bucket_count: 1 }  # Records are 1h apart and bucket span is 1h so 1 bucket is produced
+  - gt:     { jobs.0.timing_stats.minimum_bucket_processing_time_ms: 0.0 }
+  - gt:     { jobs.0.timing_stats.maximum_bucket_processing_time_ms: 0.0 }
+  - gt:     { jobs.0.timing_stats.average_bucket_processing_time_ms: 0.0 }
+  - gt:     { jobs.0.timing_stats.exponential_average_bucket_processing_time_ms: 0.0 }
```
Comment (droberts195) [inline x-pack/plugin/src/test/resources/rest-api-spec/test/ml/jobs_get_stats.yml]: It's probably safer to use `gte` on lines 104-107 and 143-146.  The jobs are only creating results for 1 bucket, and if the bucket processing time for that 1 bucket happens to round down to 0 then the `gt` condition will fail.

Comment (przemekwitek) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/ml/jobs_get_stats.yml]: Done.

File: docs/reference/ml/apis/jobcounts.asciidoc (hunk 2 above)
Comment (przemekwitek) [reply docs/reference/ml/apis/jobcounts.asciidoc]: Done.

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by przemekwitek** [comment]
run elasticsearch-ci/default-distro

**Comment by przemekwitek** [comment]
run elasticsearch-ci/1

**Review by benwtrent — APPROVED** [review]
📊

**Pull request merged**